File: mobile/test/format_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:student/utils/format.dart';
import 'package:student/widgets/status_badge.dart';

void main() {
  group('para biçimi', () {
    test('tam sayıda kuruş yazmaz, küsuratlıda iki hane gösterir', () {
      expect(formatMoney(12500), '12.500 TL');
      expect(formatMoney(12500.5), '12.500,50 TL');
    });

    test('metin girdiyi Türkçe yazımla çözer', () {
      expect(formatMoney('12.500,75'), '12.500,75 TL');
      expect(formatMoney('₺ 3.000'), '3.000 TL');
      expect(formatMoney('1,234.56'), '1.234,56 TL');
    });

    test('geçersiz değeri sıfır sayar', () {
      expect(formatMoney(null), '0 TL');
      expect(formatMoney('abc'), '0 TL');
    });

    test('işaretli biçim gelir/gideri ayırır', () {
      expect(formatMoneySigned(500), '+500 TL');
      expect(formatMoneySigned(-500), '−500 TL');
    });
  });

  group('tarih biçimi', () {
    test('gün ve ayı iki haneli yazar', () {
      expect(formatDate('2026-08-02T09:00:00'), '02.08.2026');
      expect(formatDate(DateTime(2026, 8, 2)), '02.08.2026');
    });

    test('boşluklu ISO değerini çözer', () {
      expect(formatDate('2026-08-02 09:00:00'), '02.08.2026');
    });

    test('çözülemeyen değerde yer tutucu döner', () {
      expect(formatDate(null), '—');
      expect(formatDate('bir tarih değil'), '—');
    });

    test('uzun ve kısa biçimler', () {
      expect(formatDateLong(DateTime(2026, 8, 2)), '2 Ağustos 2026');
      expect(formatMonthYear(DateTime(2026, 8, 2)), 'Ağustos 2026');
      expect(formatDayShort(DateTime(2026, 8, 2)), '02.08 Paz');
      expect(formatDateTime(DateTime(2026, 8, 2, 14, 35)), '02.08.2026 14:35');
    });
  });

  group('durum sözlüğü', () {
    test('aynı anlamı taşıyan yazımları tek etikete bağlar', () {
      expect(resolveStatus('Ödendi').label, 'Ödendi');
      expect(resolveStatus('paid').label, 'Ödendi');
      expect(resolveStatus('PAID').tone, StatusTone.success);
    });

    test('normalize anahtar Türkçe karakterleri eler', () {
      expect(normalizeStatusKey('Kısmi Ödeme'), 'kismiodeme');
      expect(normalizeStatusKey('İptal'), 'iptal');
    });

    test('bilinmeyen durumda metni korur', () {
      final custom = resolveStatus('Kendi Durumum');
      expect(custom.label, 'Kendi Durumum');
      expect(custom.tone, StatusTone.neutral);
    });

    test('masaüstüyle aynı etiketler', () {
      expect(resolveStatus('overdue').label, 'Gecikti');
      expect(resolveStatus('cancelled').label, 'İptal');
      expect(resolveStatus('Active').label, 'Aktif');
    });
  });
}

File: desktop/src/lib/format.test.js
import {
  formatDate,
  formatDateTime,
  formatDayShort,
  formatMoney,
  formatMoneySigned,
  formatNumber,
  parseMoney,
  toDate,
} from './format';

describe('formatMoney', () => {
  it('tam sayıda kuruş yazmaz, küsuratlıda iki hane gösterir', () => {
    expect(formatMoney(12500)).toBe('12.500 TL');
    expect(formatMoney(12500.5)).toBe('12.500,50 TL');
  });

  it('metin girdiyi Türkçe yazımla çözer', () => {
    expect(formatMoney('12.500,75')).toBe('12.500,75 TL');
    expect(formatMoney('₺ 3.000')).toBe('3.000 TL');
  });

  it('boş/geçersiz değeri sıfır sayar', () => {
    expect(formatMoney(null)).toBe('0 TL');
    expect(formatMoney('abc')).toBe('0 TL');
  });

  it('para birimi kodunu kısaltmaya çevirir', () => {
    expect(formatMoney(100, { currency: 'TRY' })).toBe('100 TL');
    expect(formatMoney(100, { currency: 'USD' })).toBe('100 USD');
    expect(formatMoney(100, { showCurrency: false })).toBe('100');
  });

  it('işaretli biçimde gelir/gideri ayırır', () => {
    expect(formatMoneySigned(500)).toBe('+500 TL');
    expect(formatMoneySigned(-500)).toBe('−500 TL');
    expect(formatMoneySigned(0)).toBe('0 TL');
  });
});

describe('parseMoney', () => {
  it('Türkçe ve İngilizce yazımı ayırt eder', () => {
    expect(parseMoney('1.234,56')).toBeCloseTo(1234.56);
    expect(parseMoney('1,234.56')).toBeCloseTo(1234.56);
    expect(parseMoney(1234.56)).toBeCloseTo(1234.56);
  });
});

describe('formatNumber', () => {
  it('binlik ayırıcıyı Türkçe kurala göre koyar', () => {
    expect(formatNumber(1234567)).toBe('1.234.567');
    expect(formatNumber(12.5)).toBe('12,50');
  });
});

describe('tarih biçimleri', () => {
  const iso = '2026-08-02T14:35:00Z';

  it('günü ve ayı iki haneli yazar', () => {
    // Eski `toLocaleDateString("tr-TR")` "2.8.2026" üretiyordu.
    expect(formatDate('2026-08-02T09:00:00')).toBe('02.08.2026');
  });

  it('epoch ve Date girdisini kabul eder', () => {
    const date = new Date('2026-08-02T09:00:00');
    expect(formatDate(date)).toBe('02.08.2026');
    expect(formatDate(date.getTime())).toBe('02.08.2026');
  });

  it('boşluklu ISO değerini çözer', () => {
    expect(formatDate('2026-08-02 09:00:00')).toBe('02.08.2026');
  });

  it('çözülemeyen değerde yer tutucu döner', () => {
    expect(formatDate('')).toBe('—');
    expect(formatDate('bir tarih değil')).toBe('—');
    expect(formatDate(null, '-')).toBe('-');
    expect(toDate('bir tarih değil')).toBeNull();
  });

  it('tarih+saat ve kısa gün biçimleri tutarlı', () => {
    const local = new Date(iso);
    const [datePart, timePart] = formatDateTime(local).split(' ');
    expect(datePart).toBe(formatDate(local));
    expect(timePart).toMatch(/^\d{2}:\d{2}$/);
    expect(formatDayShort(local)).toMatch(/^\d{2}\.\d{2}/);
  });
});
